### PR TITLE
Make --fail-fast configurable

### PR DESCRIPTION
**Pull request by jackscotti**
The aim of this PR is making `fail-fast` configurable, requested here: https://github.com/rspec/rspec-core/issues/1874

### CODE REVIEW

File: features/step_definitions/additional_cli_steps.rb [hunk 1]
```diff
@@ -98,7 +98,7 @@
 end
 
 Given /^I have a brand new project with no files$/ do
-  in_current_dir do
+  in_current_directory do
```
Comment (JonRowe) [inline features/step_definitions/additional_cli_steps.rb]: Are these intentional changes? Or have you just not rebased recently...

File: lib/rspec/core/example_group.rb [hunk 2]
```diff
@@ -525,7 +525,7 @@ def self.run(reporter=RSpec::Core::NullReporter)
           for_filtered_examples(reporter) { |example| example.skip_with_exception(reporter, ex) }
           true
         rescue Exception => ex
-          RSpec.world.wants_to_quit = true if fail_fast?
+          RSpec.world.wants_to_quit = true if self.fail_fast?
```
Comment (JonRowe) [inline lib/rspec/core/example_group.rb]: Uncessary change...

File: lib/rspec/core/option_parser.rb [hunk 3]
```diff
@@ -70,7 +70,18 @@ def parser(options)
           bisect_and_exit(argument)
         end
 
-        parser.on('--[no-]fail-fast', 'Abort the run on first failure.') do |value|
+        parser.on('--[no-]fail-fast[=COUNT]', 'Abort the run after a certain number of failures (1 by default).') do |argument|
```
Comment (JonRowe) [inline lib/rspec/core/option_parser.rb]: I think the `=` in this string is what forces the use of `=` on the cli, did you try dropping it?

File: features/step_definitions/additional_cli_steps.rb (hunk 1 above)
Comment (JonRowe) [inline features/step_definitions/additional_cli_steps.rb]: I think these are why the build is failing BTW, we've upgraded some Aruba things lately

File: spec/integration/filtering_spec.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 RSpec.describe 'Filtering' do
   include_context "aruba support"
-  before { clean_current_dir }
+  before { clean_current_directory }
```
Comment (JonRowe) [inline spec/integration/filtering_spec.rb]: As above I think these are accidentally changed

File: spec/integration/order_spec.rb [hunk 5]
```diff
@@ -147,7 +147,7 @@
   end
 
   context 'when a custom order is configured' do
-    after { remove_file 'spec/custom_order_spec.rb' }
+    after { remove 'spec/custom_order_spec.rb' }
```
Comment (JonRowe) [inline spec/integration/order_spec.rb]: As above I think these are accidentally changed

File: features/step_definitions/additional_cli_steps.rb (hunk 1 above)
Comment (jackscotti) [inline features/step_definitions/additional_cli_steps.rb]: This changes are part of the first commit. I have no idea why, but I had failures locally that could be only silenced by renaming these methods. The build is failing only because of these methods that have been renamed. Give me a second, I'll delete the commit and push the branch up again.

File: spec/integration/filtering_spec.rb (hunk 4 above)
Comment (jackscotti) [inline spec/integration/filtering_spec.rb]: part of the first commit

File: spec/integration/order_spec.rb (hunk 5 above)
Comment (jackscotti) [inline spec/integration/order_spec.rb]: as above

File: lib/rspec/core/option_parser.rb (hunk 3 above)
Comment (jackscotti) [inline lib/rspec/core/option_parser.rb]: I've added the "=" intentionally because if I drop it, the parser will believe that any other following string is a parameter.

ie
` --fail-fast -profile`

won't work because it will believe that `-profile` is the parameter for `--fail-fast`. Am I doing something wrong?

**Comment by jackscotti** [comment]
@JonRowe, thank you for reviewing this PR. I've removed the first commit (the one that renamed methods). The only remaining failure is the one generated by Rubocop about not being happy about the length of example_group. :disappointed:

File: lib/rspec/core/example_group.rb (hunk 2 above)
Comment (jackscotti) [inline lib/rspec/core/example_group.rb]: :+1:  going to update it

File: lib/rspec/core/option_parser.rb (hunk 3 above)
Comment (JonRowe) [inline lib/rspec/core/option_parser.rb]: Well it should be `--profile` to be considered seperate but if that's a typo the square brackets are what should make it optional

**Comment by jackscotti** [comment]
I am going to update this PR during the weekend, I think I know how to make Rubocop happy. Thank you for the feedback.

**Comment by jackscotti** [comment]
@JonRowe, I managed to get rid of Rubocop's warning by extracting the logic into its own module. The remaining failures do not appear to be related to my changes (?). Are you familiar with them? 

I am going to squash/reword commits and start to look into the parser after having understood this errors.

**Comment by JonRowe** [comment]
The last remaining build failures are due to documentation coverage dropping below 100%

File: lib/rspec/core/fail_fast_handler.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+module RSpec
+  module Core
+    module FailFastHandler
```
Comment (JonRowe) [inline lib/rspec/core/fail_fast_handler.rb]: There's no docs for this module which is what is putting off the coverage, mark them with `# @api private` if they're not intended for public usage.

**Comment by jackscotti** [comment]
I tried doing as you said but the test coverage does not change.

Without `#api private`:
screen shot 2015-09-10 at 09 51 33

With:
screen shot 2015-09-10 at 09 51 33

**Comment by samphippen** [comment]
@jackscotti  can you share the full yarddoc output for undocumented classes?

**Comment by jackscotti** [comment]
@samphippen, sorry I have been away from any electronic device for the last 2 weeks. Friday I will restart to work on this (and provide you the yarddoc).

**Comment by samphippen** [comment]
@jackscotti no problem, :+1:

**Comment by jackscotti** [comment]
Before the last commit:

screen shot 2015-10-02 at 11 10 58

After last commit:

screen shot 2015-10-02 at 11 11 14

Now the test coverage should be 100%, but Travis says it's not. Any idea?

**Comment by myronmarston** [comment]
You fixed the missing _documentation_ coverage but the build is failing due to missing _test_ coverage.  As of #1905 we enforce that rspec-core has 100% test coverage on ruby 2.1 and above. This has helped identify dead code and some bugs.  To see this locally, you'll need to run `script/rspec_with_simplecov`. That's a special binary we have to use to use simplecov with RSpec on itself because simplecov has to be loaded before all the code you are intending to check coverage of -- which means that the normal `rspec` command won't work because RSpec loads itself first.  Running that locally reveals there is missing spec coverage for [this line](https://github.com/jackscotti/rspec-core/blob/26b553f13c8c14e4e8fcf24642e7e0ac3097462b/lib/rspec/core/fail_fast_handler.rb#L12) and [these lines](https://github.com/jackscotti/rspec-core/blob/26b553f13c8c14e4e8fcf24642e7e0ac3097462b/lib/rspec/core/option_parser.rb#L79-L83).

File: lib/rspec/core/example_group.rb [hunk 7]
```diff
@@ -26,7 +26,7 @@ module Core
     # your examples defined in {MemoizedHelpers} and {Pending}.
     class ExampleGroup
       extend Hooks
-
+      extend FailFastHandler
```
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: Since `FailFastHandler` is completely private and doesn't have any APIs we want to expose to users, there's no need for the methods to be added to `ExampleGroup` like this.  After all, it's not a property of the example group class whether or not the failures number has been met.  I'd prefer it if you just made the methods exposed by `FailFastHandler` module methods and called them as e.g. `FailFastHandler.failures_number_met?` w/o extending the module.

File: lib/rspec/core/example_group.rb [hunk 8]
```diff
@@ -557,7 +557,9 @@ def self.run_examples(reporter)
           instance = new(example.inspect_output)
           set_ivars(instance, before_context_ivars)
           succeeded = example.run(instance, reporter)
-          RSpec.world.wants_to_quit = true if fail_fast? && !succeeded
+          if !succeeded && fail_fast?
+            RSpec.world.wants_to_quit = true if failures_number_met?(reporter.failed_examples.count)
+          end
```
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: It seems odd to do this:

``` ruby
if some_condition
  do_something if some_other_condition
end
```

...rather than:

``` ruby
if some_condition && some_other_condtion
  do_something
end
```

Mind refactoring to the latter?

File: lib/rspec/core/example_group.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: Actually, seeing `failures_number_met?(reporter.failed_examples.count)` below makes me think that these method defs should perhaps move to the reporter: then it can just be `reporter.fail_fast_limit_met?`.  After all, the reporter holds the state of `reporter.failed_examples.count` already and it has a reference to the configuration instance.  We try to limit how much we access our global state (e.g. `RSpec.configuration` and `RSpec.world`) so it's nice to use instances that objects have already been passed such as reporter already having a config instance.

File: spec/rspec/core/example_group_spec.rb [hunk 9]
```diff
@@ -1372,6 +1373,37 @@ def extract_execution_results(group)
...
+        let(:config)   { Configuration.new }
+        let(:reporter) { Reporter.new config }
+
+        it "does not run examples after 3 failed examples" do
+          examples_run = []
+          self.group.example('example 1') { examples_run << self }
+          self.group.example('example 2') { examples_run << self; fail; }
+          self.group.example('example 3') { examples_run << self; fail; }
+          self.group.example('example 4') { examples_run << self; fail; }
+          self.group.example('example 5') { examples_run << self }
```
Comment (JonRowe) [inline spec/rspec/core/example_group_spec.rb]: You don't need `self.` here... rename the local instead to avoid the ruby 2.2 warning

File: lib/rspec/core/example_group.rb (hunk 7 above)
Comment (jackscotti) [inline lib/rspec/core/example_group.rb]: The way of implementing it you're suggesting sounds better, I will address it as soon as I can. Thank you for the great feedback and apologise for the noise, I am pretty unfamiliar with this codebase.

**Comment by jackscotti** [comment]
Alright, seems like we managed to get Simplecov happy. Thank you again for all the feedback, I will address it next time I'll work again on this. :)

File: lib/rspec/core/example_group.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: No apology needed.  Thanks for contributing!  RSpec is better because of contributions like yours.

**Comment by jackscotti** [comment]
I changed laptop and simplecov does not want to work, would you mind telling me what line is not being covered?

File: lib/rspec/core/example_group.rb [hunk 10]
```diff
@@ -557,11 +556,17 @@ def self.run_examples(reporter)
           instance = new(example.inspect_output)
           set_ivars(instance, before_context_ivars)
           succeeded = example.run(instance, reporter)
-          RSpec.world.wants_to_quit = true if fail_fast? && !succeeded
+          if !succeeded && fail_fast?
+            RSpec.world.wants_to_quit = true if reporter.fail_fast_limit_met?
+          end
```
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: I'd still like to see the two `if` statements here combined into one.  Nesting them seems messier and unnecessary to me.

File: lib/rspec/core/example_group.rb [hunk 11]
```diff
@@ -557,11 +556,17 @@ def self.run_examples(reporter)
...
           succeeded
         end.all?
       end
 
+      def self.fail_fast?
+        RSpec.configuration.fail_fast
+      end
+
```
Comment (myronmarston) [inline lib/rspec/core/example_group.rb]: Any reason you moved this?  As far as I can see this isn't needed...and generally it's nice not to move code unless that's the point of the PR or commit.  You mind reverting it and rebasing it out of the commits in your PR?

**Comment by jackscotti** [comment]
@myronmarston I tried rebasing on master but `Failure/Error: expect(error_messages.compact).to be_well_formed License.md has spaces on the EOL on lines 4, 5, 6` does not go away. Is this something that has been fixed?

**Comment by jackscotti** [comment]
@myronmarston , I have one error I do not understand and one of code not covered. That line of code hasn't been introduced by me, why do you think Simplecov is picking it up only now? And, are you familiar with the error?

**Comment by myronmarston** [comment]
> regarding the error, I just found this: https://bugs.ruby-lang.org/issues/10661 and understood why @JonRowe asked me to rename the variable.

Yeah, the Ruby 2.2.0 warning is _extremely_ annoying, particularly because of RSpec's `let` feature.  It looks like the latest build is still getting a failure from that -- this time from `reporter`.  I noticed your change in the last commit, where you renamed `group` to `the_group` in a bunch of places.  Usually to fix this you want your local variables to be named differently from your `let` definitions but you changed them both to `the_group` which I would have thought would have not fixed the warning (it looks like it did, though -- I'm a bit confused as to how!).  So I think you just need to make a similar change to `reporter`.  Sorry for the trouble :(.  I really wish that warning had never been introduced in Ruby 2.2.0.

**Comment by jackscotti** [comment]
WOOOOHOOOO!! All green!
@myronmarston , thank you for the good explanation, I was finding really hard to silence those warnings. I rebased that commit, now it has a much cleaner diff.

To note that if I address @JonRowe comment about removing `self.` the warnings come back.

This PR is now ready for review, I will squash/rename/reword the commits ASAP. 
Thank you all for the great feedback and support. :)

File: lib/rspec/core/option_parser.rb (hunk 3 above)
Comment (JonRowe) [inline lib/rspec/core/option_parser.rb]: Changing this string to `'--[no-]fail-fast [count]'` drops the need for the `=`, I double checked.

File: features/configuration/fail_fast.feature [hunk 12]
```diff
@@ -1,19 +1,13 @@
 Feature: fail fast
 
-  Use the `fail_fast` option to tell RSpec to abort the run on first failure:
+  Use the `fail_fast` option to tell RSpec to abort the run on after N failures:
 
-  ```ruby
-  RSpec.configure { |c| c.fail_fast = true }
-  ```
```
Comment (JonRowe) [inline features/configuration/fail_fast.feature]: This is documentation and should be restored (although you can add a comment that true means 1)

File: features/command_line/fail_fast.feature [hunk 13]
```diff
@@ -3,17 +3,29 @@ Feature: `--fail-fast` option
   Use the `--fail-fast` option to tell RSpec to stop running the test suite on
   the first failed test.
 
-  You may also specify `--no-fail-fast` to turn it off (default behaviour).
+  You may add a parameter to tell RSpec to stop running the test suite after N
+  failed tests, for example: `--fail-fast=3`.
```
Comment (JonRowe) [inline features/command_line/fail_fast.feature]: If you make my suggested change you'll have to drop these `=` of course, but you should also mention the default is `1`.

File: features/configuration/fail_fast.feature (hunk 12 above)
Comment (JonRowe) [inline features/configuration/fail_fast.feature]: Do we still have an example of just a bare `--fail-fast`?

File: lib/rspec/core/example_group.rb [hunk 14]
```diff
@@ -576,7 +578,7 @@ def self.for_filtered_examples(reporter, &block)
 
       # @private
       def self.fail_fast?
-        RSpec.configuration.fail_fast?
+        RSpec.configuration.fail_fast
```
Comment (JonRowe) [inline lib/rspec/core/example_group.rb]: Why this change?

**Comment by JonRowe** [comment]
>To note that if I address @JonRowe comment about removing self. the warnings come back.

>This PR is now ready for review, I will squash/rename/reword the commits ASAP. 

Just a few final tweaks, and yep if you could squash the commits that'd be great :)

File: lib/rspec/core/option_parser.rb (hunk 3 above)
Comment (myronmarston) [inline lib/rspec/core/option_parser.rb]: @JonRowe when I added the bisect feature and wanted to support `--bisect verbose` I couldn't get that to work and had to add the `=` in the middle.  I couldn't figure out why.  Do you want to take a look at that?

File: features/command_line/fail_fast.feature (hunk 13 above)
Comment (jackscotti) [inline features/command_line/fail_fast.feature]: I won't be able to concatenate any other command if I drop the `=`. Maybe I am doing something wrong but if I drop it, it will see anything following `fail-fast` as a parameter for `fail-fast` itself, therefore if I add another command it will error. 

For example we have `--bisect` that requires an `=`.
What do you suggest?

File: features/configuration/fail_fast.feature (hunk 12 above)
Comment (jackscotti) [inline features/configuration/fail_fast.feature]: here: https://github.com/rspec/rspec-core/pull/2065/files#diff-87d87ffcd904e2f75195937563248991L20

**Comment by jackscotti** [comment]
As you can see if I drop the `=` tests fail.

**Comment by jackscotti** [comment]
Is there anything I can do to get this PR merged?

**Comment by samphippen** [comment]
@jackscotti when was the last time you rebased against master? I definitely saw something about aruba flying by in my inbox, so that might solve the problem.

**Comment by JonRowe** [comment]
I'd still like to solve the issue with the CLI, tests pass for me and it works but cucumber doesn't seem to like it

**Comment by jackscotti** [comment]
Thanks for getting back to me so quickly! 

Rebased on master, now the error is changed. Don't really undertand what is complaining about? It seems to be related to Aruba and not to the code I've introduced.

**Comment by myronmarston** [comment]
The failure isn't related to aruba other than the fact that it's a cucumber scenario and we use aruba in our cucumber scenarios and "aruba" is in the directory name that gets used for the scenario.  The failing scenario is the [`--fail-fast 3`](https://github.com/rspec/rspec-core/pull/2065/files#diff-87d87ffcd904e2f75195937563248991R37) one that you added in this PR.  It looks like it's interpreting the 3 in `--fail-fast 3` as a file name and not as an argument to `--fail-fast`.  Probably caused by 46eb102bcb5d9b17655a04ac846e5ac44f8b8b0c?

**Comment by jackscotti** [comment]
Yes, you are right. 

We are back to the original error then. What should the parser accept? 
I can delete that commig if you are happy with the `=`.

**Comment by myronmarston** [comment]
I have no idea how to get it to work without the `=`.  I ran into the same thing with `bisect`.  @JonRowe seems to think it's possible to get it to work but for now remove the commit that took the `=` out and Jon can always do a follow up PR showing us how to make it work without `=` if he can figure it out.

**Comment by myronmarston** [comment]
BTW, before merging it would also be nice to not have 19 commits (including some merge commits).  Do you mind rebasing and squashing those out so that the PR has a small number of logical commits?

If that's difficult for you (it's one of those git things that's tricky to do for the first time) let me know and I can push up what you've got to a branch so we can get it merged from there.

Also, sorry for the long delay on this.  We've done a poor job at helping you get this ready :(.

**Comment by jackscotti** [comment]
Really no need to be sorry, I know you've all got a lot to do!!

I am going to rebase/squash/reword the commits and let you know when I think I am done, so you can tell me whether the final result is good or not.

Thank you for the great feedback and patience.

**Comment by jackscotti** [comment]
Updated and squashed the commits. 
The weird Ruby 2.2 warning appeared again. :sob: I think the fix makes sense to be on its own commit rather being squashed in others. What do you recon?

**Comment by myronmarston** [comment]
Thanks, @jackscotti!
